cli/pkg/kctrl/cmd/package/release/helm_openapi_schema_gen.go

```diff
@@ -0,0 +1,362 @@
+// Copyright 2022 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package release
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+
+	kcdatav1alpha1 "github.com/vmware-tanzu/carvel-kapp-controller/pkg/apiserver/apis/datapackaging/v1alpha1"
+	yaml2 "gopkg.in/yaml.v2"
+	yaml3 "gopkg.in/yaml.v3"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/yaml"
+)
+
+// keys used when generating an OpenAPI Document
+const (
+	typeKey         = "type"
+	formatKey      = "format"
+	descriptionKey = "description"
+	itemsKey      = "items"
+	propertiesKey = "properties"
+	defaultKey    = "default"
+)
+
+var propOrder = map[string]int{
+	typeKey:        1,
+	formatKey:      2,
+	descriptionKey: 3,
+	itemsKey:       4,
+	propertiesKey:  5,
+	defaultKey:     6,
+}
+
+const (
+	objectVal = "object"
+	arrayVal  = "array"
+	floatVal  = "float"
+)
+
+// Yaml tags that differentiate the type of scalar object in the node
+const (
+	nullTag  = "!!null"
+	boolTag  = "!!bool"
+	intTag   = "!!int"
+	floatTag = "!!float"
+)
+
+type MapItem struct {
+	Key   interface{}
+	Value interface{}
+}
+
+type Map struct {
+	Items []*MapItem
+}
+
+func (m *Map) Iterate(iterFunc func(k, v interface{})) {
+	for _, item := range m.Items {
+		iterFunc(item.Key, item.Value)
+	}
+}
+
+type openAPIKeys []*MapItem
+
+func (o openAPIKeys) Len() int {
+	return len(o)
+}
+
+func (o openAPIKeys) Less(i, j int) bool {
+	return propOrder[o[i].Key.(string)] < propOrder[o[j].Key.(string)]
+}
+
+func (o openAPIKeys) Swap(i, j int) {
+	o[i], o[j] = o[j], o[i]
+}
+
+type HelmValuesSchemaGen struct {
+	// dirPath to the helm chart directory
+	dirPath string
+}
+
+func NewHelmValuesSchemaGen(path string) HelmValuesSchemaGen {
+	return HelmValuesSchemaGen{path}
+}
+
+func (h HelmValuesSchemaGen) Schema() (*kcdatav1alpha1.ValuesSchema, error) {
+	fileData, err := h.readValuesFile()
+	if err != nil {
+		return nil, err
+	}
+	if len(fileData) == 0 {
+		return &kcdatav1alpha1.ValuesSchema{
+			OpenAPIv3: runtime.RawExtension{Raw: nil},
+		}, nil
+	}
+
+	var document yaml3.Node
+	err = yaml3.Unmarshal(fileData, &document)
+	if err != nil {
+		return nil, err
+	}
+	if document.Kind == 0 {
```

Comment:
I am not sure if the value of `Kind` can be zero post marshalling (loss like valid values are powers of 2 (here)[https://pkg.go.dev/gopkg.in/yaml.v3#Kind])

However, since we are doing nothing if that's the case, I would say we should do this if `IsZero()`
```suggestion
	if document.IsZero() {
```

Reply:
Nice, adopted